### PR TITLE
feat: add --cache_mb and --cache_percentage flag

### PR DESCRIPTION
**Pull request by NamanJain8**
Add --cache_mv and --cache-percentage flags for `dgraph alpha` and `dgraph zero`
Format:
`pblocksize:pindexsize:windexsize:postingListsize`  --> alpha
`blockcache:indexCache` --> zero

Will update the description and code needs to be cleaned up too.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6271)

### CODE REVIEW

**Review by jarifibrahim — CHANGES_REQUESTED** [review]
Got some comments.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-:-MF_bTnj4y4RfimLPntM:b17ku4n)* status: 0 of 11 files reviewed, 12 unresolved discussions (waiting on @manishrjain, @martinmr, @NamanJain8, and @vvbalaji-dgraph)

___
*[go.mod, line 9 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF__RCj5HYWG0IMG3s1:-MF__RCj5HYWG0IMG3s2:bs1nolr) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/go.mod#L9)):*
> ```XML
> 	contrib.go.opencensus.io/exporter/prometheus v0.1.0
> 	github.com/99designs/gqlgen v0.12.2
> 	github.com/AndreasBriese/bbloom v0.0.0-20190825152654-46b345b51c96 // indirect
> ```

Run `go mod tidy` and it should remove this.
___
*[dgraph/cmd/alpha/run.go, line 209 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_ac5cDIhiSIW0XJAk:-MF_ac5cDIhiSIW0XJAl:b5la44p) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L209)):*
> ```Go
> 
> 	// Add cache flags
> 	flag.Int64("cache_mb", 0, "Total size of cache (in MB) to be used in dgraph")
> ```

to be used in alpha
___
*[dgraph/cmd/alpha/run.go, line 210 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF__WVsBOMwP9zeuw7Z:-MF__WVsBOMwP9zeuw7_:b-htam3f) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L210)):*
> ```Go
> 	// Add cache flags
> 	flag.Int64("cache_mb", 0, "Total size of cache (in MB) to be used in dgraph")
> 	flag.String("cache-percentage", "30:30:20:20", "Cache percentages for various caches (pblocksize:pindexsize:windexsize:postingListsize)")
> ```

100 characters.

cache-percentage => cache_percentage.

Add a detailed explanation about the ratios. `pblocksize` would make no sense to the user.
___
*[dgraph/cmd/alpha/run.go, line 592 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF__i2z1ZYKTzoIUsdu:-MF__i3--jDWrGc1kxoe:bll8lzv) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L592)):*
> ```Go
> 	x.AssertTruef(totalCache >= 0, "ERROR: Cache size must be non-negative")
> 	if len(cp) != 4 {
> 		log.Fatalf("ERROR: cache percentage format is pblocksize:pindexsize:windexsize:postingListsize")
> ```

Here as well. Instead of calling it `pblockSize` let's call it `PstoreBlockCache` and `PstoreIndexCache` , etc. We want user to be able to understand what these flags actually mean.
___
*[dgraph/cmd/alpha/run.go, line 610 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_a18CCMHGTJKizBhF:-MF_a18D5rcYyaFhQtqV:b-w5jg6q) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L610)):*
> ```Go
> 
> 	if percentSum != 100 {
> 		log.Fatalf("ERROR: cache percentage does not sum up to 100")
> ```

Add the percentages here. w+x+y+z is not equal to 100.
___
*[dgraph/cmd/alpha/run.go, line 615 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_aEdh-OIAvycMZlWz:-MF_aEdh-OIAvycMZlX-:bjd445q) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L615)):*
> ```Go
> 	pIndexSize := (cachePercent[1] * (totalCache << 20)) / 100
> 	wIndexSize := (cachePercent[2] * (totalCache << 20)) / 100
> 	postingListSize := (cachePercent[3] * (totalCache << 20)) / 100
> ```

All these are cache sizes. Let's call them cache sizes.
___
*[dgraph/cmd/debug/run.go, line 787 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_aLRu96MzAzEqmspv:-MF_aLRu96MzAzEqmspw:bqyf5sl) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/debug/run.go#L787)):*
> ```Go
> 	}
> 	x.Check(err)
> 	posting.Init(db, 0)
> ```

Add a comment `// Not using posting list cache`
___
*[dgraph/cmd/zero/run.go, line 112 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_a_ok20hvVaofb0-n:-MF_a_ok20hvVaofb0-o:b-mpxxoj) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L112)):*
> ```Go
> 
> 	// Add cache flags
> 	flag.Int64("cache_mb", 0, "Total size of cache (in MB) to be used in dgraph")
> ```

to be used in zero
___
*[dgraph/cmd/zero/run.go, line 113 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_aizs0oVyjKQmYfkA:-MF_aizs0oVyjKQmYfkB:b-ilr2hq) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L113)):*
> ```Go
> 	// Add cache flags
> 	flag.Int64("cache_mb", 0, "Total size of cache (in MB) to be used in dgraph")
> 	flag.String("cache-percentage", "75:25", "Cache percentages for various caches (blockcache:indexCache)")
> ```

cache_percentage

Zero doesn't have an index cache right now. We should set this to `100:0`right now. We don't need caches always.
___
*[dgraph/cmd/zero/run.go, line 230 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_atKZ-po9kCHHtZ8y:-MF_atKZ-po9kCHHtZ8z:badhrvw) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L230)):*
> ```Go
> 	}
> 
> 	cp := strings.Split(opts.cachePercentage, ":")
> ```

You could move the code below to `x.go` file (or if we have a util file in `x`). The code is duplicated.
___
*[dgraph/cmd/zero/run.go, line 242 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_b4cEDxpZqrnile2g:-MF_b4cEDxpZqrnile2h:bewrvte) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L242)):*
> ```Go
> 		x, err := strconv.Atoi(percent)
> 		if err != nil {
> 			log.Fatalf("ERROR: ERROR: unable to parse cache percentages")
> ```

Print the `percent` here.
___
*[dgraph/cmd/zero/run.go, line 245 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_b7f42dc9LcZFCGwD:-MF_b7f42dc9LcZFCGwE:b252cmu) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L245)):*
> ```Go
> 		}
> 		if x < 0 {
> 			log.Fatalf("ERROR: cache percentage cannot be negative")
> ```

Print the `percent` here as well.

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6271)* status: 7 of 12 files reviewed, 17 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[go.mod, line 9 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF__RCj5HYWG0IMG3s1:-MFeH--83AAvXT7ObrMJ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/go.mod#L9)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Run `go mod tidy` and it should remove this.
</blockquote></details>

Done.
___
*[dgraph/cmd/alpha/run.go, line 209 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_ac5cDIhiSIW0XJAk:-MFeH3zs80yoGfiq4VLa:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L209)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

to be used in alpha
</blockquote></details>

Done.
___
*[dgraph/cmd/alpha/run.go, line 210 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF__WVsBOMwP9zeuw7Z:-MFeH61Y0m_bsmeYHCEb:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L210)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

100 characters.

cache-percentage => cache_percentage.

Add a detailed explanation about the ratios. `pblocksize` would make no sense to the user.
</blockquote></details>

Done.
___
*[dgraph/cmd/alpha/run.go, line 592 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF__i2z1ZYKTzoIUsdu:-MFeHL0k4GsVUIigL8aE:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L592)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Here as well. Instead of calling it `pblockSize` let's call it `PstoreBlockCache` and `PstoreIndexCache` , etc. We want user to be able to understand what these flags actually mean.
</blockquote></details>

Done.
___
*[dgraph/cmd/alpha/run.go, line 610 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_a18CCMHGTJKizBhF:-MFeHOP48qRD6w8-PKpt:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L610)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Add the percentages here. w+x+y+z is not equal to 100.
</blockquote></details>

Done.
___
*[dgraph/cmd/alpha/run.go, line 615 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_aEdh-OIAvycMZlWz:-MFeHQeu3ml_m_ljNWo1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/alpha/run.go#L615)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

All these are cache sizes. Let's call them cache sizes.
</blockquote></details>

Done.
___
*[dgraph/cmd/alpha/run.go, line 210 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_xqHF4bRaX9cX5Paq:-MFeHC6G8AzA8oxod5eX:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/acb6bfac8eb2f7606f30e3133c1a6c7ab24835bb/dgraph/cmd/alpha/run.go#L210)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Order it in the biggest to smallest expected caches.

posting list cache : p store block cache : p store index cache : w store block cache : w store index cache

Use commas.

`0,65,25,0,10`
</blockquote></details>

Done.
___
*[dgraph/cmd/debug/run.go, line 787 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_aLRu96MzAzEqmspv:-MFeHSbKAVrfM6N68NUd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/debug/run.go#L787)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Add a comment `// Not using posting list cache`
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/run.go, line 112 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_a_ok20hvVaofb0-n:-MFeHTmR6LHoQs3dnCa1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L112)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

to be used in zero
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/run.go, line 113 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_aizs0oVyjKQmYfkA:-MFeHXWR9bPtaq1_-bZY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L113)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars. Use commas.
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/run.go, line 230 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_atKZ-po9kCHHtZ8y:-MFeHYYe0KIuw7R1FTFF:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L230)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

You could move the code below to `x.go` file (or if we have a util file in `x`). The code is duplicated.
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/run.go, line 242 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_b4cEDxpZqrnile2g:-MFeHa7W-lKC7HUMj5R1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L242)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Print the `percent` here.
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/run.go, line 245 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MF_b7f42dc9LcZFCGwD:-MFeH_ff4JEee1p_KiPh:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/1cf5d3b2dc9d2e0fc82e39fc8bdbeacae5253a12/dgraph/cmd/zero/run.go#L245)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Print the `percent` here as well.
</blockquote></details>

Done.
___
*[x/x.go, line 1106 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFa0yMs7mZYVqcMdx6Q:-MFeHbBX871QTr1E20c3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/acb6bfac8eb2f7606f30e3133c1a6c7ab24835bb/x/x.go#L1106)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

numExpected int

don't need format.
</blockquote></details>

Done.
___
*[x/x.go, line 1107 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFa0lD90lQqQF7R17lM:-MFeHbb14VFVgxgnDQSw:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/acb6bfac8eb2f7606f30e3133c1a6c7ab24835bb/x/x.go#L1107)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

comma.
</blockquote></details>

Done.
___
*[x/x.go, line 1109 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFa0uWZ1Wu9K-tphK7R:-MFeHc3E-1v3CkoAGFBr:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/acb6bfac8eb2f7606f30e3133c1a6c7ab24835bb/x/x.go#L1109)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

numExpected int
</blockquote></details>

Done.
___
*[x/x.go, line 1118 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFa19is2DFqfjEzQSJy:-MFeHePr8nmagMUx5mei:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/acb6bfac8eb2f7606f30e3133c1a6c7ab24835bb/x/x.go#L1118)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

don't need to do fatals here. Return the error and let the caller deal with this.
</blockquote></details>

Done.

**Review by jarifibrahim — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6271)* status: 4 of 12 files reviewed, 10 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @martinmr, @NamanJain8, and @vvbalaji-dgraph)

___
*[dgraph/cmd/bulk/reduce.go, line 133 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFeJds_CA3ZFyb65GjZ:-MFeJds_CA3ZFyb65Gj_:b-7i7pig) ([raw file](https://github.com/dgraph-io/dgraph/blob/84673e4457c9151d75e931b1d2317db508c3afe0/dgraph/cmd/bulk/reduce.go#L133)):*
> ```Go
> 	opt := badger.DefaultOptions(dir).WithSyncWrites(false).
> 		WithTableLoadingMode(bo.MemoryMap).WithValueThreshold(1 << 10 /* 1 KB */).
> 		WithLogger(nil).WithBlockCacheSize(1 << 20).
> ```

This isn't related to this PR but this cache is too small. 
___
*[worker/config.go, line 68 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFeLqb973XqZChIDJOF:-MFeLqb973XqZChIDJOG:b8g40tk) ([raw file](https://github.com/dgraph-io/dgraph/blob/84673e4457c9151d75e931b1d2317db508c3afe0/worker/config.go#L68)):*
> ```Go
> 	// PIndexCacheSize is the size of index cache for pstore
> 	PIndexCacheSize int64
> 	// PBlockCacheSize is the size of block cache for wstore
> ```

wblockcachesize
___
*[worker/config.go, line 70 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFeLsMrDzdF9lITeN9x:-MFeLsMrDzdF9lITeN9y:bs50sqr) ([raw file](https://github.com/dgraph-io/dgraph/blob/84673e4457c9151d75e931b1d2317db508c3afe0/worker/config.go#L70)):*
> ```Go
> 	// PBlockCacheSize is the size of block cache for wstore
> 	WBlockCacheSize int64
> 	// PIndexCacheSize is the size of index cache for wstore
> ```

windexcachesize
___
*[x/x.go, line 1109 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFeNQbhA9Mmo9aEVvAi:-MFeNQbhA9Mmo9aEVvAj:b-dqyxyv) ([raw file](https://github.com/dgraph-io/dgraph/blob/84673e4457c9151d75e931b1d2317db508c3afe0/x/x.go#L1109)):*
> ```Go
> 	// Sanity checks
> 	if len(cp) != numExpected {
> 		return nil, errors.Errorf("ERROR: expected %d cache percentages, got %d", numExpected, len(cp))
> ```

100 chars

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6271)* status: 4 of 12 files reviewed, 10 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @martinmr, @NamanJain8, and @vvbalaji-dgraph)

___
*[worker/config.go, line 68 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFeLqb973XqZChIDJOF:-MFeQms91OranPYweeJQ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/84673e4457c9151d75e931b1d2317db508c3afe0/worker/config.go#L68)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

wblockcachesize
</blockquote></details>

Done.
___
*[worker/config.go, line 70 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFeLsMrDzdF9lITeN9x:-MFeQoTy2vOICdwl3Cmv:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/84673e4457c9151d75e931b1d2317db508c3afe0/worker/config.go#L70)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

windexcachesize
</blockquote></details>

Done.
___
*[x/x.go, line 1109 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6271#-MFeNQbhA9Mmo9aEVvAi:-MFeQslr0V3VCSz23egQ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/84673e4457c9151d75e931b1d2317db508c3afe0/x/x.go#L1109)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

100 chars
</blockquote></details>

Done.

File: dgraph/cmd/bulk/reduce.go [hunk 1]
```diff
@@ -130,7 +130,7 @@ func (r *reducer) createBadgerInternal(dir string, compression bool) *badger.DB
 
 	opt := badger.DefaultOptions(dir).WithSyncWrites(false).
 		WithTableLoadingMode(bo.MemoryMap).WithValueThreshold(1 << 10 /* 1 KB */).
-		WithLogger(nil).WithMaxCacheSize(1 << 20).
+		WithLogger(nil).WithBlockCacheSize(1 << 20).
```
Comment (parasssh) [inline dgraph/cmd/bulk/reduce.go:133]: shouldn't we use the flag configured instead of 1 <<  20 ?

File: x/x.go [hunk 2]
```diff
@@ -1099,3 +1099,35 @@ func DeepCopyJsonArray(a []interface{}) []interface{} {
...
+	var cachePercent []int64
+	percentSum := 0
+	for _, percent := range cp {
+		x, err := strconv.Atoi(percent)
+		if err != nil {
+			return nil, errors.Errorf("ERROR: unable to parse cache percentage(%s)", percent)
+		}
+		if x < 0 {
+			return nil, errors.Errorf("ERROR: cache percentage(%s) cannot be negative", percent)
+		}
```
Comment (parasssh) [inline x/x.go:1122]: add check that x > 100 is invalid too.

Comment (NamanJain8) [reply x/x.go:1122]: This case will be handled in sum percentages.

File: dgraph/cmd/bulk/reduce.go (hunk 1 above)
Comment (NamanJain8) [reply dgraph/cmd/bulk/reduce.go:133]: I think we should. That would require adding a flag in `dgraph bulk`. @jarifibrahim  What do you say?

Comment (jarifibrahim) [reply dgraph/cmd/bulk/reduce.go:133]: Bulk loader shouldn't even have a cache. I'll create a ticket for this. This Pr doesn't need to block on this.

Comment (jarifibrahim) [reply dgraph/cmd/bulk/reduce.go:133]: I see that those tmpDBs are being read. We need a cache here but it has to be configurable. I have created DGRAPH-2352

**Pull request merged**